test/framework/deployment_helpers.go

```diff
@@ -179,51 +187,129 @@ type watchPodLogsInput struct {
+	return &watchPodLogsEventHandler{
+		ctx:         ctx,
+		input:       input,
+		selector:    selector,
+		startedPods: sync.Map{},
+	}
+}
+
+func (eh *watchPodLogsEventHandler) OnAdd(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	eh.streamPodLogs(*pod)
+}
+
+func (eh *watchPodLogsEventHandler) OnUpdate(_, newObj interface{}) {
+	pod := newObj.(*corev1.Pod)
+	eh.streamPodLogs(*pod)
+}
 
-				podLogs, err := input.ClientSet.CoreV1().Pods(input.Namespace).GetLogs(pod.Name, opts).Stream(ctx)
+func (eh *watchPodLogsEventHandler) OnDelete(_ interface{}) {}
+
+func (eh *watchPodLogsEventHandler) streamPodLogs(pod corev1.Pod) {
+	if pod.GetNamespace() != eh.input.Namespace {
+		return
+	}
+	if !eh.selector.Matches(labels.Set(pod.GetLabels())) {
+		return
+	}
+	if pod.Status.Phase != corev1.PodRunning {
+		return
+	}
+	if _, loaded := eh.startedPods.LoadOrStore(pod.GetUID(), struct{}{}); loaded {
+		return
+	}
+
+	for _, container := range pod.Spec.Containers {
+		log.Logf("Creating log watcher for controller %s, pod %s, container %s", klog.KRef(eh.input.Namespace, eh.input.DeploymentName), pod.Name, container.Name)
+
+		// Create log metadata file.
+		logMetadataFile := filepath.Clean(path.Join(eh.input.LogPath, eh.input.DeploymentName, pod.Name, container.Name+"-log-metadata.json"))
+		Expect(os.MkdirAll(filepath.Dir(logMetadataFile), 0750)).To(Succeed())
+
+		metadata := logMetadata{
+			Job:       eh.input.Namespace + "/" + eh.input.DeploymentName,
+			Namespace: eh.input.Namespace,
+			App:       eh.input.DeploymentName,
+			Pod:       pod.Name,
+			Container: container.Name,
+			NodeName:  pod.Spec.NodeName,
+			Stream:    "stderr",
+		}
+		metadataBytes, err := json.Marshal(&metadata)
+		Expect(err).To(BeNil())
+		Expect(os.WriteFile(logMetadataFile, metadataBytes, 0600)).To(Succeed())
+
+		// Watch each container's logs in a goroutine so we can stream them all concurrently.
+		go func(pod corev1.Pod, container corev1.Container) {
+			defer GinkgoRecover()
+
+			logFile := filepath.Clean(path.Join(eh.input.LogPath, eh.input.DeploymentName, pod.Name, container.Name+".log"))
+			Expect(os.MkdirAll(filepath.Dir(logFile), 0750)).To(Succeed())
+
+			f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+			Expect(err).NotTo(HaveOccurred())
+			defer f.Close()
+
+			opts := &corev1.PodLogOptions{
+				Container: container.Name,
+				Follow:    true,
+			}
+
+			// Retry streaming the logs of the pods unless ctx.Done() or if the pod does not exist anymore.
+			err = wait.PollInfiniteWithContext(eh.ctx, 2*time.Second, func(ctx context.Context) (done bool, err error) {
+				// Wait for pod to be in running state
+				actual, err := eh.input.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
 				if err != nil {
+					if apierrors.IsNotFound(err) {
+						return true, nil
+					}
+					// Failing to get the pod should not cause the test to fail
```

Comment:
nit: This whole thing doesn't fail the test in any case as it's a separate go routine and we only log in the end
(let's update the comment)